src/runtime/testing/sequence.ts

```diff
@@ -0,0 +1,594 @@
+    this.changes[id] = changes.map(({input, inputFn, output, variable}) => 
+      ({input, inputFn, output, variable, key: id})
+    );
+  }
+
+  /**
+   * set an end invariant for a given sensor.
+   * 
+   * @param id The sensor to set an end invariant on
+   * @param test a function that takes the final sensor value and asserts properties
+   */
+  setEndInvariant(id: string, test: (t: any) => void) {
+    this.sensors.get(id).endInvariants.push(test);
+  }
+
+  private resetVariables() {
+    for (const variable of this.variables.values()) {
+      if (variable.initialFn) {
+        variable.value = variable.initialFn();
+      } else {
+        variable.value = variable.initialValue;
+      }
+    }
+  }
+
+  private resetResults() {
+    for (const input of this.inputs.values()) {
+      input.results = [];
+    }
+  }
+
+  private async awaitResults() {
+    for (const input of this.inputs.values()) {
+      input.results = await Promise.all(input.results);
+    }
+  }
+
+  private setupOutputs(obj) {
+
+    for (const output of this.outputs.values()) {
+      const responseChecker = value => {
+        this.interleavingLog.push('->');
+        this.interleavingLog.push(value);
+        this.interleavingLog.push('\n');
+        if (output.response.onOutput) {
+          try {
+            output.response.onOutput(value);
+          } catch (e) {
+            console.log(...this.interleavingLog);
+            throw e;
+          }
+        }
+        return output.value;
+      };
+
+      if (output.response.type === ExpectedResponse.Defer) {
+        output.value = output.response.default;
+      }
+
+      const parts = output.name.split('.');
+      let theObject = obj;
+      for (let i = 0; i < parts.length - 1; i++) {
+        theObject = theObject[parts[i]];
+      }
+
+      if (output.behavior === SequenceOutput.Replace) {
+        theObject[parts[parts.length - 1]] = responseChecker;
+      } else if (output.behavior === SequenceOutput.Register) {
+        const result = theObject[parts[parts.length - 1]](responseChecker);
+        if (output.variable) {
+          this.setVariable(output.variable, result);
+        }
+      }
+    }  
+  }
+
+    /*
+   * length: total number of stages.
+   * amounts: an array of inputs, with the number of stages in each input.
+   *
+   * Generate all the permutations of choice of the next stage in each input
+   * until all stages are exhausted.
+   * 
+   * Note that rather than maintaining indices into possibly empty inputs,
+   * this function removes inputs as they become empty, and thus produces
+   * interleavings that act on dynamically reassigned input numbers.
+   * 
+   * For example, if there are two inputs [a,b] and [c,d] then the possible
+   * interleavings are: [a,b,c,d], [a,c,b,d], [a,c,d,b], [c,a,b,d], [c,a,d,b], [c,d,a,b]
+   * These will be expressed at this level as:
+   * [0, 0, 0, 0], [0, 1, 0, 0], [0, 1, 1, 0], [1, 0, 0, 0], [1, 0, 1, 0], [1, 1, 0, 0]
```

Comment:
Wat.

I understand the list of interleavings, but not the encoding.

Reply:
Don't worry too much about it, it's an internal implementation detail

Reply:
Except if someone else needs to fix/update this function, it'd be useful to ensure the details are clear. :)

Reply:
[0, 0, 0, 0] means: 
-take from the first non-empty list ([a, b])
-take from the first non-empty list ([b])
-take from the first non-empty list ([c,d])
-take from the first non-empty list ([d])
-> [a,b,c,d]